src/sagemaker/image.py

```diff
@@ -0,0 +1,446 @@
+                'ports': [
+                    '8080:8080'
+                ]
+            })
+
+        return host_config
+
+    def _optml_volumes(self, host, subdirs, single_model_dir=False):
+        """
+        It takes a folder with the necessary files for training and creates a list of opt volumes that
+        the Container needs to start.
+        If args.single_model_dir is True, all the hosts will point the opt/ml/model subdir to the first container.
+        That is useful for distributed training, so all the containers can read and write the same checkpoints.
+
+        :param opt_root_folder: root folder with the contents to be mapped to the container
+        :param host: host name of the container
+        :param subdirs: list of subdirs that will be mapped. Example: ['input', 'output', 'model']
+        :return:
+        """
+        volumes = []
+
+        # If it is single mode dir we want to map the same model dir and share between hosts
+        if single_model_dir:
+            host_dir = os.path.join(self.container_root, 'algo-1/model')
+            volume = Volume(host_dir, '/opt/ml/model')
+            volumes.append(volume)
+        else:
+            # else we want to add model to the list of subdirs so it will be created for each container.
+            subdirs.add('model')
+
+        for subdir in subdirs:
+            host_dir = os.path.join(self.container_root, host, subdir)
+            container_dir = '/opt/ml/{}'.format(subdir)
+            volume = Volume(host_dir, container_dir)
+            volumes.append(volume)
+
+        return volumes
+
+
+class _Container(object):
+    def __init__(self, tmpdir, command, startup_delay=5):
+        self.command = command
+        self.compose_file = os.path.join(tmpdir, DOCKER_COMPOSE_FILENAME)
+        self.startup_delay = startup_delay
+        self._process = None
+
+    def up(self):
+        self._process = Popen(self.command)
+        sleep(self.startup_delay)
+
+    def down(self):
+        self._process.terminate()
+        _cleanup()
+
+
+class Volume(object):
+
+    def __init__(self, host_dir, container_dir=None, channel=None):
+        # that is necessary because docker cannot mount mac temp folders.
+        if not container_dir and not channel:
+            raise ValueError('Either container_dir or channel must be declared.')
+
+        if container_dir and channel:
+            raise ValueError('container_dir and channel cannot be declared together.')
+
+        self.container_dir = container_dir if container_dir else os.path.join('/opt/ml/input/data', channel)
+
+        self.host_dir = os.path.join('/private', host_dir) if host_dir.startswith('/var') else host_dir
+
+        self.map = '{}:{}'.format(self.host_dir, self.container_dir)
+
+
+def _cleanup():
+    _chain_docker_cmds('docker images -f dangling=true -q', 'docker rmi -f')
+    _check_output('docker network prune -f'.split(' '))
+
+
+def _chain_docker_cmds(cmd, cmd2):
```

Comment:
Can you add some developer docs to this.

Also - is there any reason why this doesn't take an arg list, why just two commands?

Reply:
Got rid of this function. again this was carry over that I didn't touch but I got rid of it now, it doesn't make much sense and it doesn't 'chain' docker commands.